### PR TITLE
Replace data providers with doctrine extensions

### PR DESCRIPTION
**Pull request by paullla**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/CountryCollectionExtensionSpec.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        QueryNameGeneratorInterface $queryNameGenerator,
+        ChannelInterface $channel
+    ):void {
+        $queryBuilder->getRootAliases()->willReturn(['o']);
+
+        $userContext->getUser()->willReturn($admin);
+        $admin->getRoles()->willReturn(['ROLE_API_ACCESS']);
+
+        $queryBuilder->andWhere('o..id in :countries')->shouldNotBeCalled();
+        $queryBuilder->setParameter('countries', $channel->getCountries())->shouldNotBeCalled();
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/CountryCollectionExtensionSpec.php:46]: In situation like this we could use argument helper to define, that any value will break this condition
```suggestion
        $queryBuilder->setParameter('countries', Argument::any())->shouldNotBeCalled();
```

File: src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/CountryCollectionExtensionSpec.php [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+use Sylius\Component\Core\Model\ChannelInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+final class CountryCollectionExtensionSpec extends ObjectBehavior
+{
+    function let(UserContextInterface $userContext): void
+    {
+        $this->beConstructedWith($userContext);
+    }
+    function it_does_not_apply_conditions_for_admin(
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/CountryCollectionExtensionSpec.php:33]:
```suggestion

    function it_does_not_apply_conditions_for_admin(
```

Comment (lchrusciel) [reply src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/CountryCollectionExtensionSpec.php:33]: We are also missing check for lacking of channel context

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/CountryCollectionExtension.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    private UserContextInterface $userContext;
+
+    public function __construct(UserContextInterface $userContext)
+    {
+        $this->userContext = $userContext;
+    }
+
+    public function applyToCollection(
+        QueryBuilder $queryBuilder,
+        QueryNameGeneratorInterface $queryNameGenerator,
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/CountryCollectionExtension.php:36]: Do you have any experience/recommendations about `queryNameGenerator` usage?

File: src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/CurrencyCollectionExtensionSpec.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+use Sylius\Component\Currency\Model\CurrencyInterface;
+use Symfony\Component\HttpFoundation\Request;
+
+final class CurrencyCollectionExtensionSpec extends ObjectBehavior
+{
+    function let(UserContextInterface $userContext): void
+    {
+        $this->beConstructedWith($userContext);
+    }
+    function it_does_not_apply_conditions_for_admin(
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/CurrencyCollectionExtensionSpec.php:33]:
```suggestion

    function it_does_not_apply_conditions_for_admin(
```
+ test for lack of channel in context

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/TaxonCollectionExtension.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Util\QueryNameGeneratorInterface;
+use Doctrine\ORM\QueryBuilder;
+use Sylius\Bundle\ApiBundle\Context\UserContextInterface;
+use Sylius\Bundle\ApiBundle\Serializer\ContextKeys;
+use Sylius\Component\Core\Model\AdminUserInterface;
+use Sylius\Component\Taxonomy\Model\TaxonInterface;
+use Webmozart\Assert\Assert;
+
+/** @experimental */
+final class TaxonCollectionExtension implements ContextAwareQueryCollectionExtensionInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/TaxonCollectionExtension.php:26]: we are missing spec for this class

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/AddressGetMethodItemExtension.php [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        string $resourceClass,
+        array $identifiers,
+        string $operationName = null,
+        array $context = []
+    ) {
+        if (!is_a($resourceClass, AddressInterface::class, true)) {
+            return;
+        }
+
+        $rootAlias = $queryBuilder->getRootAliases()[0];
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/AddressGetMethodItemExtension.php:52]: Could be part of `applyToItemForGetMethod` method

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/AddressGetMethodItemExtension.php [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+use Sylius\Component\Core\Model\AddressInterface;
+use Sylius\Component\Core\Model\AdminUserInterface;
+use Sylius\Component\Core\Model\CustomerInterface;
+use Sylius\Component\Core\Model\ShopUserInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException;
+use Symfony\Component\Security\Core\User\UserInterface;
+
+/** @experimental */
+final class AddressGetMethodItemExtension implements QueryItemExtensionInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/AddressGetMethodItemExtension.php:31]: As we are using this extension for all methods, lets remove this words to reduce possible misunderstanding
```suggestion
final class AddressItemExtension implements QueryItemExtensionInterface
```

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/ProductItemExtension.php [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+use Sylius\Bundle\ApiBundle\Context\UserContextInterface;
+use Sylius\Bundle\ApiBundle\Serializer\ContextKeys;
+use Sylius\Component\Core\Model\AdminUserInterface;
+use Sylius\Component\Core\Model\ChannelInterface;
+use Sylius\Component\Core\Model\ProductInterface;
+use Sylius\Component\Core\Model\ShopUserInterface;
+use Webmozart\Assert\Assert;
+
+/** @experimental */
+class ProductItemExtension implements QueryItemExtensionInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/ProductItemExtension.php:28]:
```suggestion
final class ProductItemExtension implements QueryItemExtensionInterface
```

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/CountryCollectionExtension.php [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        if (!is_a($resourceClass, CountryInterface::class, true)) {
+            return;
+        }
+
+        $user = $this->userContext->getUser();
+        if ($user instanceof AdminUserInterface && in_array('ROLE_API_ACCESS', $user->getRoles(), true)) {
+            return;
+        }
+
+        $channel = $context[ContextKeys::CHANNEL] ?? null;
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/CountryCollectionExtension.php:50]: To make it coherent among extensions, perhaps we should thrown an exception when channel is not present int context

**Comment by soyuka** [comment]
Hi, what's the reasoning behind this? In my opinion the provider is the place to fetch data. Maybe that by decorating our provider you should find an in-between solution? On top of that the "Extension" part is quite doctrine-specific and its implementation is, in my opinion, not the cleanest. I'd really suggest not to use it for the use case I see in this PR.

**Comment by paullla** [comment]
@soyuka when using data providers we are not able to use filters and doctrine extensions for that endpoint. Is there a way to make it happen with data providers?

**Comment by lchrusciel** [comment]
Hey @soyuka, I had the impression that with data providers we are short-cutting api platform logic. When one would like to take advantage of limiting functionalities, filters etc. These changes seems to bring our solution closer to yours

**Comment by soyuka** [comment]
I read the code again and indeed I understand. My bad the code is fine, the PR should be named differently though ^^. The goal here is to add a doctrine extension point (for example add a WHERE clause on the country if needed). This is better indeed!

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/CountryCollectionExtension.php [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+        $user = $this->userContext->getUser();
+        if ($user instanceof AdminUserInterface && in_array('ROLE_API_ACCESS', $user->getRoles(), true)) {
+            return;
+        }
+
+        Assert::keyExists($context, ContextKeys::CHANNEL);
+        $channel = $context[ContextKeys::CHANNEL];
+
+        if ($channel !== null && $channel->getCountries()->count() > 0) {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/CountryCollectionExtension.php:54]: As we've introduced assert for channel existence, we can remove this check 
```suggestion
        if ($channel->getCountries()->count() > 0) {
```

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/CurrencyCollectionExtension.php [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+        $user = $this->userContext->getUser();
+        if ($user instanceof AdminUserInterface && in_array('ROLE_API_ACCESS', $user->getRoles(), true)) {
+            return;
+        }
+
+        Assert::keyExists($context, ContextKeys::CHANNEL);
+        $channel = $context[ContextKeys::CHANNEL];
+
+        if ($channel !== null && $channel->getCurrencies()->count() > 0) {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/CurrencyCollectionExtension.php:54]:
```suggestion
        if ($channel->getCurrencies()->count() > 0) {
```

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/LocaleCollectionExtension.php [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+        $user = $this->userContext->getUser();
+        if ($user instanceof AdminUserInterface && in_array('ROLE_API_ACCESS', $user->getRoles(), true)) {
+            return;
+        }
+
+        Assert::keyExists($context, ContextKeys::CHANNEL);
+        $channel = $context[ContextKeys::CHANNEL];
+
+        if ($channel !== null && $channel->getLocales()->count() > 0) {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/LocaleCollectionExtension.php:54]:
```suggestion
        if ($channel->getLocales()->count() > 0) {
```

File: src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/CountryCollectionExtensionSpec.php [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    function let(UserContextInterface $userContext): void
+    {
+        $this->beConstructedWith($userContext);
+    }
+
+    function it_throws_an_exception_if_context_has_not_channel(
+        QueryBuilder $queryBuilder,
+        QueryNameGeneratorInterface $queryNameGenerator
+    ): void
+    {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/CountryCollectionExtensionSpec.php:39]:
```suggestion
    ): void {
```

File: src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/CurrencyCollectionExtensionSpec.php [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+    function let(UserContextInterface $userContext): void
+    {
+        $this->beConstructedWith($userContext);
+    }
+
+    function it_throws_an_exception_if_context_has_not_channel(
+        QueryBuilder $queryBuilder,
+        QueryNameGeneratorInterface $queryNameGenerator
+    ): void
+    {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/CurrencyCollectionExtensionSpec.php:39]:
```suggestion
    ): void {
```

File: src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/LocaleCollectionExtensionSpec.php [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+    function let(UserContextInterface $userContext): void
+    {
+        $this->beConstructedWith($userContext);
+    }
+
+    function it_throws_an_exception_if_context_has_not_channel(
+        QueryBuilder $queryBuilder,
+        QueryNameGeneratorInterface $queryNameGenerator
+    ): void
+    {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/LocaleCollectionExtensionSpec.php:38]:
```suggestion
    ): void {
```

File: src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/TaxonCollectionExtensionSpec.php [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+    function let(UserContextInterface $userContext): void
+    {
+        $this->beConstructedWith($userContext);
+    }
+
+    function it_throws_an_exception_if_context_has_not_channel(
+        QueryBuilder $queryBuilder,
+        QueryNameGeneratorInterface $queryNameGenerator
+    ): void
+    {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryCollectionExtension/TaxonCollectionExtensionSpec.php:39]:
```suggestion
    ): void {
```

File: src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryItemExtension/AddressItemExtensionSpec.php [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        $this->beConstructedWith($userContext);
+    }
+
+    function it_does_not_apply_conditions_for_admin(
+        UserContextInterface $userContext,
+        QueryBuilder $queryBuilder,
+        AdminUserInterface $admin,
+        QueryNameGeneratorInterface $queryNameGenerator
+    ):void {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/Doctrine/QueryItemExtension/AddressItemExtensionSpec.php:40]:
```suggestion
    ): void {
```
The same should be done in 7 more places

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/AddressItemExtension.php [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+        if ($user === null) {
+            throw new MissingTokenException('JWT Token not found');
+        }
+
+        $this->applyToItemForGetMethod($user, $queryBuilder);
+    }
+
+    private function applyToItemForGetMethod(?UserInterface $user, QueryBuilder $queryBuilder): void
+    {
+
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/AddressItemExtension.php:63]:
```suggestion
```

**Review by lchrusciel — APPROVED** [review]

**Comment by lchrusciel** [comment]
Thank you, Paula! :1st_place_medal:

**Pull request merged**